### PR TITLE
[Test Fix #3424] PointInTimeOperationTest.listAllPits_positive

### LINKED ISSUE #3424: [Test] Reenable disable test case PointInTimeOperationTest.listAllPits_positive
### Description
Looks like the call to the cluster `cleanUpPits` is returning before all of the PITs have actually been deleted on the cluster.  

### Repo
1. Go to `src/integrationTest/java/org/opensearch/security/PointInTimeOperationTest.java`
2. Find the `@Ignore` entry on the test named `listAllPits_positive`
3. Delete that attribute
4. Run the tests with `./gradlew integrationTest --tests org.opensearch.security.PointInTimeOperationTest.listAllPits_positive`

### Exit Criteria
- [ ] Ensure that PITs have been cleaned up in a deterministic way - note this could be a product change too
- [ ] Run the test at least 10 times and see no failures - [how to do this](https://github.com/opensearch-project/security/pull/3388/commits/a714cf5cd07c0fa624965ab0b1a7c34e5143b75e)

### PR DESCRIPTION
**Pull request by EduardoCorazon**
### Description
This PR is meant to address the issue of cleanUpPits() returning before all of the PITs have actually been deleted on the cluster as referenced in issue #3424.

[Describe what this change achieves]
* Category: Test fix
* Why these changes are required? : The file PointInTimeOperationTest.java has functions listAllPits_positive() and deleteAllPits_positive() calling cleanUpPits(). However, it was suggested that cleanUpPits() was returning before all of the PITs actually got deleted. Upon testing, the call to delete all pits in `restHighLevelClient.deleteAllPits(DEFAULT)` yields the prevalent error:
```
java.io.IOException: Unable to parse response body for Response{requestLine=DELETE /_search/point_in_time/_all HTTP/1.1, host=https://127.0.0.1:47210, response=HTTP/2.0 200 OK}
 at __randomizedtesting.SeedInfo.seed([14B732A7AAEADD96:E3D6D13BF1B12B63]:0) ... Caused by: com.fasterxml.jackson.core.io.JsonEOFException: Unexpected end-of-input: was expecting closing '"' for name
 at [Source: (ByteArrayInputStream); line: 1, column: 12]
```
* What is the old behavior before changes and new behavior after changes? : The old behavior as mentioned above, resulted in failed integration tests using gradlew and vs code debuggers. The new changes fix this issue by fetching information about existing PIT searches before attempting to delete them. It's possible this solutions works either because  there might be some internal synchronization or update to the PIT search state when we call `getAllPits` or it could be a race condition whereby there is an attempt to delete PITs before internal operations have completed.

### Issues Resolved
#3424 

Is this a backport? If so, please add backport PR # and/or commits #

### Testing
Integration tests with gradlew were run on both listAllPits_positive() and the entire class PointInTimeOperationtest(). These tests were set to run at least 10 times by configuring ci.yml as outlined here: [how to do this](https://github.com/opensearch-project/security/pull/3388/commits/a714cf5cd07c0fa624965ab0b1a7c34e5143b75e)

### Exit Criteria
- [x] Ensure that PITs have been cleaned up in a deterministic way - note this could be a product change too
- [x] Run the test at least 10 times and see no failures - [how to do this](https://github.com/opensearch-project/security/pull/3388/commits/a714cf5cd07c0fa624965ab0b1a7c34e5143b75e)

### Check List
- [x] New functionality includes testing
- [x] New functionality has been documented
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by EduardoCorazon** [comment]
Currently investigating why solution works & properly implementing the fix. Will @ when ready for full review.

**Comment by peternied** [comment]
I'm going to switch this to draft so folks get notified when you are ready for a review.

image

File: src/integrationTest/java/org/opensearch/security/CrossClusterSearchTests.java [hunk 1]
```diff
@@ -454,4 +454,4 @@ public void shouldLackAccessToSpecificField() throws IOException {
             assertThat(response, searchHitContainsFieldWithValue(0, FIELD_GENRE, GENRE_ROCK));
         }
     }
-}
+}
```
Comment (DarshitChanpura) [inline src/integrationTest/java/org/opensearch/security/CrossClusterSearchTests.java:457]: nit: this should be reverted.

**Comment by EduardoCorazon** [comment]
@DarshitChanpura I've marked this PR as ready, the original problem for cleanUpPits has been addressed. I've reverted all accidental changes from CrossClusterSearchTests.java but I'm still getting conflicts. I was wondering why this might be as there were no changes to the file outside the reverted accidental change?

**Comment by EduardoCorazon** [comment]
I also deeply apologize for the delayed response regarding this matter.

**Comment by cwperks** [comment]
Thank you @EduardoCorazon. Can you explain how adding `restHighLevelClient.getAllPits(DEFAULT);` resolves the issue with cleaning up the PITs? What is the response body when the IOException occurs?

File: src/integrationTest/java/org/opensearch/security/PointInTimeOperationTest.java [hunk 2]
```diff
@@ -413,10 +410,17 @@ private String createPitForIndices(String... indices) throws IOException {
     /**
     * Deletes all PITs.
     */
+
     public void cleanUpPits() throws IOException {
+
         try (RestHighLevelClient restHighLevelClient = cluster.getRestHighLevelClient(ADMIN_USER)) {
+
+            restHighLevelClient.getAllPits(DEFAULT);
```
Comment (DarshitChanpura) [inline src/integrationTest/java/org/opensearch/security/PointInTimeOperationTest.java:418]: qq, why add this?

**Review by DarshitChanpura — COMMENTED** [review]
@EduardoCorazon Seems like there are conflicts, would you please address them?

**Comment by EduardoCorazon** [comment]
I think there might be a synchronization issue that's affecting the behavior of cleanUpPits() and it could be why the test is flaky. Adding `restHighLevelClient.getAllPits(DEFAULT)` before attempting to delete all PITs (even if you declare .getAllPits(DEFAULT) as an object that will never be used) seemingly fixes the flaky test. It could be that the `getAllPits` method might be triggering some internal update that stabilizes all PITs before deleting them.

This can be verified by running `./gradlew integrationTest --tests org.opensearch.security.PointInTimeOperationTest.listAllPits_positive` and if you include `restHighLevelClient.getAllPits(DEFAULT)` the test is a success but if it's commented it out, the test fails.

File: src/integrationTest/java/org/opensearch/security/PointInTimeOperationTest.java (hunk 2 above)
Comment (EduardoCorazon) [reply src/integrationTest/java/org/opensearch/security/PointInTimeOperationTest.java:418]: This seemingly fixes the test flakiness. I've found that even just having the declaration of `restHighLevelClient.getAllPits(DEFAULT)` before attempting to delete all PITS makes the `listAllPits_positive` integration test succeed as opposed to failing

**Comment by EduardoCorazon** [comment]
Also, @cwperks do you know how I could push out the raw HTTP response as standard output? I've exhausted my attempts at attempting to do such given the current libraries. I apologize for the simple question.

File: src/integrationTest/java/org/opensearch/security/PointInTimeOperationTest.java (hunk 2 above)
Comment (DarshitChanpura) [reply src/integrationTest/java/org/opensearch/security/PointInTimeOperationTest.java:418]: @EduardoCorazon Have we verified the cause of this?

**Comment by peternied** [comment]
Doesn't look like this is being worked on anymore, going to close this out.  Please feel free to reopen if you re-engage on this @EduardoCorazon

**Pull request closed without merging**